Lab-4/patrik_svensson_lab_function_approximation.ipynb

```diff
@@ -0,0 +1,286 @@
+        "$a_{ii} = \\frac{h_i}{3} + \\frac{h_{i + 1}}{3}$\n",
+        "\n",
+        "According to lecture notes (9.15):\n",
+        "\n",
+        "$a_{ii+1} = \\frac{h_{i+1}}{6}$\n",
+        "\n",
+        "Accoding to lecture notes (9.16):\n",
+        "\n",
+        "$a_{ii-1} = \\frac{h_i}{6}$\n",
+        "\n",
+        "Accoding to lecture notes (9.17):\n",
+        "\n",
+        "$b_i = (f,\\phi_i) = \\int_{0}^{1} f(x)\\phi_i(x)dx$ which can be further expanded to:\n",
+        "$b_i = f(x_i)(\\frac{h_i}{2} + \\frac{h_{i+1}}{2})$ according to (1.74) in *The Finite Element Method: Theory, Implementation, and Practice* by Mats G. Larsson, Fredrik Bengzon.\n",
+        "\n",
+        "To calculate the approximation of the function that we approximated with $L_2$ projection:\n",
+        "\n",
+        "$P_hf(x) = \\sum_{j=0}^{m+1}\\alpha_j\\phi_j(x)$\n",
+        "\n",
+        "We can plug in the $\\alpha$ values that we have obtained. The $\\phi(x)$ values can be obtained from (9.7) in the lecture notes.\n",
+        "\n",
+        "$\\phi_i(x) = \\lambda_{k,1}(x)$ if $x \\in I_k$\n",
+        "\n",
+        "$\\phi_i(x) = \\lambda_{k+1,1}(x)$ if $x \\in I_{k + 1}$\n",
+        "\n",
+        "$\\phi_i(x) = 0$ if $x \\neq I_k \\cup I_{k + 1}$"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "GKAKgzyjs5ZU",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def L2Projection1D(function, mesh):\n",
+        "  A = np.zeros((mesh.shape[0], mesh.shape[0]))\n",
+        "  b = np.zeros(mesh.shape[0])\n",
+        "\n",
+        "  for i in range(A.shape[0]):\n",
+        "    if(i == (A.shape[0] - 1)):\n",
+        "      interval_size = abs(mesh[i] - mesh[i - 1])\n",
+        "      A[i][i] = interval_size / 3\n",
+        "      A[i][i-1] = interval_size / 6\n",
+        "    elif (i == 0):\n",
+        "      next_interval_size = abs(mesh[i] - mesh[i + 1])\n",
+        "      A[i][i] = next_interval_size / 3\n",
+        "      A[i][i + 1] = next_interval_size / 6\n",
+        "\n",
+        "    else:\n",
+        "      interval_size = abs(mesh[i] - mesh[i - 1]) \n",
+        "      next_interval_size = abs(mesh[i] - mesh[i + 1])\n",
+        "      A[i][i] = interval_size / 3 + next_interval_size / 3\n",
+        "      A[i][i + 1] = next_interval_size / 6\n",
+        "      A[i][i - 1] = interval_size / 6\n",
+        "\n",
+        "\n",
+        "  for i in range(A.shape[0] - 1):\n",
+        "    interval_size = 0\n",
+        "    function_val = function(mesh[i])\n",
+        "    if(i == (A.shape[0] - 1)):\n",
+        "      interval_size = abs(mesh[i] - mesh[i-1])\n",
+        "    elif (i == 0):\n",
+        "      interval_size = abs(mesh[i] - mesh[i + 1])\n",
+        "    else:\n",
+        "      interval_size = abs(mesh[i] - mesh[i-1]) + abs(mesh[i] - mesh[i + 1])\n",
+        "\n",
+        "    b[i] = interval_size * function_val\n",
```

Comment:
Your calculation of b is incorrect:

1. Consider revisiting your own definition and approximation of bi.
2. Are all bi calculated correctly? Consider printing the b when you have finished it.